### ISSUE TITLE
bug: tabSize affects indentation

### ISSUE THREAD

**Issue by cblp** [microsoft/vscode#10339]
- VSCode Version: Code 1.4.0 (6276dcb0ae497766056b4c09ea75be1d76a8b679, 2016-08-04T16:49:32.489Z)
- OS Version: Linux x64 4.4.0-34-generic

`tabSize` setting has two independent functions:
1. Size of tabulation character.
2. Size of indentation.

The latter is really useful. But its connectedness with the former makes it unusable. In one file tabulation character size may be 8 and indentation size may be 4. How can I control these two values with one setting?

**Comment by rmunn** [comment]
>  In one file tabulation character size may be 8 and indentation size may be 4. How can I control these two values with one setting?

@cblp Have you tried the EditorConfig plugin? It's written for exactly this scenario. See http://editorconfig.org/ for more details.

**Comment by cblp** [comment]
Yet another plugin and yet another configuration file for a single option? Okay :(

**Comment by cblp** [comment]
No, EditorConfig doesn't fix the issue. When I configure it to
```ini
[*]
indent_size = 2
tab_width = 8
```
it changes **both** indentation size **and** tab size to 2.

I think, EditorConfig can't work properly because of this issue.

**Comment by gcp** [comment]
This bug also makes proper handling of mode lines an issue, see for example https://github.com/ctlajoie/vscode-modelines/issues/1.

**Comment by marktucker** [comment]
This issue makes it impossible for me to use VS Code as an editor on our existing library of code (which uses vim soft tabs of 4 spaces to indent, but a tab size of 8). With this fixed I could replace all my terminal windows, Vim, and Visual Studio with VS Code... Pretty please?

**Comment by pcdummy** [comment]
I also realy would welcome this to be fixed.

**Comment by alexandrudima** [comment]
Thanks to @dlech in the PR #62079, there is now a new setting `indentSize` that controls how many spaces the indentation is.

This is by default configured to be equal to `editor.tabSize`, but the two settings can now be independently configured.

**Comment by octref** [comment]
This does not seem to work.

I have the following setting:

```json
{
  "editor.tabSize": 8,
  "editor.indentSize": 2
}
```

Tab is still rendered as 4 spaces wide:

![image](https://user-images.githubusercontent.com/4033249/53378416-56d14900-391a-11e9-9599-6c71dbd09640.png)

Status bar shows Spaces: 4:

![image](https://user-images.githubusercontent.com/4033249/53378429-605ab100-391a-11e9-9a95-bf5986181714.png)

Meanwhile, tab to insert spaces only insert 2 spaces:

![tab](https://user-images.githubusercontent.com/4033249/53378555-d828db80-391a-11e9-8b1e-ca36e6023467.gif)

I have tried disabling all extensions (including Vim and editor config) and all other settings, still does not work for me.

**Comment by octref** [comment]
@dlech Would you mind trying it out in https://code.visualstudio.com/insiders/ and see what went wrong?

**Comment by dlech** [comment]
I found a suspicions change: https://github.com/Microsoft/vscode/pull/62079/commits/b164459bf868366d9825c7e9ad5118ebb6a16f91#r260115575

I'll try to have a look at the unit tests later this week to make sure we are covering all the proper cases to catch issues like this.

**Comment by alexandrudima** [comment]
@octref You must also configure `editor.detectIndentation: false`

**Comment by octref** [comment]
Thanks for the explanation. Verified that it works. Sorry @dlech for the false alarm.

**Comment by octref** [comment]
@alexandrudima Had a sync with @rebornix and @Tyriar, we didn't like the setting name because with the new setting, `vscode.FormattingOptions.tabSize` takes VS Code's `tabSize`, which is now "Size of tabulation character" but is should be "Size of indentation".

```json
{
  "editor.detectIndentation": false,
  "editor.tabSize": 2,
  "editor.indentSize": 4
}
```

Formatter should receive `4` instead of `2`. Currently it receives `2`.

We think a better name might be `editor.visualTabSize`, since:

- This setting is clear to only affect visual representation of a hard tab.
- No implicit inheritance of `tabSize` from `indentSize`, and no `indentSize` depends on `detectIndentation` to be false. Both behaviors are confusing.
- We don't need to make changes to formatter API, because `vscode.FormattingOptions.tabSize` is there and already carry the meaning of "Size of indentation".

**Comment by alexandrudima** [comment]
@octref `editor.tabSize` exists since > 5 years, so, no, we cannot really change the name from `editor.tabSize` to `editor.visualTabSize`.

- the fact that the formatter receives 2 is a bug, we should fix that.
- `indentSize` inherits from `tabSize`. The default for `indentSize` is `"tabSize"`.
- `indentSize` behaves like `tabSize`, i.e. they both will be detected from the opened file if `editor.detectIndentation` is true.
- yes, we need to pass `indentSize` as the value for `tabSize` when making API calls to formatters.

ping @jrieken

**Comment by Tyriar** [comment]
@alexandrudima the problem this appears to solve for me is allowing users to specify how they want tabs to be viewed, even if this is set in workspace settings. I like viewing TS with 2 spaces for example, it would be nice if I could view the vscode code base in 2 tabs, regardless of whether detectIndentation is true or false.

That's why "editor.visualTabSize" came up, I would expect the jsdoc for it to read something like this:

> The visual size of a tab character on screen, when this is not specified editor.tabSize will be used.

I can currently achieve something very close to this in insiders with these user settings:

```json
  "[typescript]": {
    "editor.tabSize": 2
  },
  "[javascript]": {
    "editor.tabSize": 2
  }
```

But this has a couple of issues:

- This could mess potentially mess with another project's tabs if insertSpaces is true
- This could be overridden by a workspace setting (but shouldn't be as it's a user preference)

One of the nice things about the `editor.visualTabSize` approach is that it's a very simple change that is always active if there are tab characters on the page. The current solution doesn't really allow users to state their preference and have it work in all workspaces since tabSize and indentSize only do anything if detectIndentation is off.

**Comment by alexandrudima** [comment]
OK, I'm pulling the plug on this one.

I cannot give this discussion the attention it deserves so late in the endgame and there is no reason to rush with a decision, especially since we have survived without 2 distinct settings for so many years.

**Comment by dlech** [comment]
Do we have metrics on the current usage of the related settings here? I am curious because I always set `"editor.detectIndentation": false` globally on all of my VS Code installations. I have found that the detection rarely does the right thing (because much of the source code I am looking at is not perfectly formatted).

And it seems worth reiterating the problem we are actually trying to fix here. There exist source code files such as those in [HarfBuzz](https://github.com/harfbuzz/harfbuzz/blob/45149eb34f9735b5d690a2a7956adb42b938c8d9/src/hb-font.cc#L139-L147) where the chosen code style is to indent using 2 spaces... but every instance of 8 spaces is replaced with a tab character.

```C
    { // 4 spaces
      if (!font->get_nominal_glyph (*first_unicode, first_glyph)) // 6 spaces
	return i; // tab character
```

Currently, in order to view these files in VS Code with proper indentation, you have to set `tabSize` to 8. However, this causes automatic indentation to not work properly because it always indents 8 spaces when it should only indent 2. So you end up pressing backspace 6 times on every indent or just use a different text editor. Likewise, if you set `tabSize` to 2, the file is just not readable.

**Comment by rebornix** [comment]
@dlech 

> we have metrics on the current usage of the related settings here

For the source code you provided above, the metrics of `editor.detectIndentation` and `tabSize` are below

| editor.detectIndentation | editor.tabSize | effective soft tab width | hard tab width |
|--------------------------|----------------|--------------------------|----------------|
| true                     | 2              | 2                        | 2              |
| true                     | 4              | 2                        | 2              |
| false                    | 2              | 2                        | 2              |
| false                     | 4              | 4                        | 4              |

The /effective soft tab width/ works as expected and to appeal to HarfBuzz's coding style, we can either turn on `editor.detectIndentation` or set `editor.tabSize` to 2, then we are sure that every time when users press Tab, two spaces are inserted into the editor. Unless the file is poorly formatted.

> but every instance of 8 spaces is replaced with a tab character
> ...
> Currently, in order to view these files in VS Code with proper indentation, you have to set `tabSize` to 8

If I understand correctly, HarfBuzz assumes that hard tab is always rendered as 8 space wide so it replaces all 8 spaces to hard tabs by purpose. IMO `editor.visualTabSize` can simply solve this problem as it won't touch any logic of indenting, formatting, etc but only affect how a hard tab is being rendered.

**Comment by dlech** [comment]
RE: naming of the settings... here is some research on "prior art"

Visual Studio has "Tab Size" and "Indent Size"

https://stackoverflow.com/a/9089823/1976323

Eclipse has "Indentation Size" and "Tab Size"

https://superuser.com/a/462223/324575

Sublime Text only has "tabSize" ([similar feature request](https://forum.sublimetext.com/t/separate-tab-stops-indent-size-and-character-settings/13085/15))

https://www.sublimetext.com/docs/indentation

Atom has "tabLength" ([similar feature request](https://github.com/atom/atom/issues/12931))

`https://flight-manual.atom.io/using-atom/sections/basic-customization/#configuration-key-reference

Vim has "shiftwidth" and "tabstop"

https://vi.stackexchange.com/a/7977

The EditorConfig spec has "tab_width" and "indent_size"

https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#widely-supported-by-editors

**Comment by dlech** [comment]
I understand the problem with my approach better now. With the current "tabSize" setting meaning both the number of columns to indent and the display size of the tab character, the most important aspect is the indentation. Since my changes introduced a new "indentSize" setting to control indentation, this can easily lead to broken projects when a user has globally set the "indentSize" and a project has set the "tabSize" (expecting the old behavior of controlling both aspects).

So, what is really needed is a new setting that controls the number of columns used for the tab stop for displaying tab characters if and only if a project is indenting using spaces. Such a setting should have no effect when `"insertSpaces": false`.

Some open questions are:

1. The name. "visualTabSize" has been suggested. My personal opinion is that it would be nice to try to be consistent with other major editors. This isn't really possible because of the fact that the most important aspect of the existing "tabSize" setting is the indentation, so the meaning can't be the same as Visual Studio and Eclipse. However, I kind of like "tabstop" from vim since the visual width of a tab character could actually be less than 8 characters (or whatever the setting is set to).

2. The default behavior. If this new setting only applies when `"insertSpaces": true`, I don't see much harm in setting the default value to `8`. This matches the majority of the legacy code that this feature is meant to address. So, although this would change the default behavior of VS Code, I think most users might think this is an improvement.

**Comment by octref** [comment]
> Such a setting should have no effect when "insertSpaces": false.

If the naming is `visualTabSize`, I think it applies even when `insertSpaces` is true, because setting `insertSpaces` true does not remove the existing hard-tabs in files, and those hard-tabs should be rendered at `visualTabSize` column wide.

**Comment by Wintaru** [comment]
I'd like to vote for this to be fixed as well. I tried the EditorConfig extension, and when I reload the window it indeed "fixes" the issue (shows tabs with the correct spaces I put in .editorconfig, 8) but as soon as I click on another file and then back to the original, Code resets the tab width back to 4 (which weirdly, is not what I have in Code's default configuration for Editor:Tab Size, I put 8 there too).

**Comment by trusktr** [comment]
If I understand correctly and am in the right place, this is important for develop experience. For example, some people really like to view code with indentation being at the width of four spaces. Regardless if the underlying code is actually using 2-space indentation, the developer wants to view it with 4-space indentation (equivalent of tab characters set to 4 spaces wide) to make the development experience better. However, VS Code does not allow this, instead, VS Code always modifies the underlying code to match the developers viewing preference, which is undersirable when collaborating with other people who may rely on the 2-space indentation.

Basically what we need is to control how the code is actually formatted, and to control _how we view_ the code, such that if we open the code in a simple plain text editor we would see 2-space indentation, but in VS Code (with appropriate settings) we would see it as 4-space-tab-like only for developer convenience.

Is this what people are asking for?

**Comment by marktucker** [comment]
I don't want to speak for anyone else, but what I want is much simpler than this, I think. Actual "space" characters should be sacrosanct, not considered part of some "indentation" that the editor tries to interpret. All I want is that a tab character be allowed to move to the next columne that is a multiple of 8, which at the same time have my "indent size" for automatic code formatting set to 4 spaces (real space characters, no tab characters).

For me, the use case is dealing with an old code base filled with a mix of tabs characters and space characters (but with a consistent indent step size of 4 columns, populated as either space characters or size-8 tab characters) looks right in the editor. And I can edit these files without reformatting/editing every single line (automatically or not) without polluting my version control system with useless edits to the code spacing.

**Comment by Wintaru** [comment]
What @marktucker said. I would like to be able to set tab characters to 8 spaces but when I hit the tab key want it to indent 4 spaces. My use case is the same as his.

Before anyone says "well fix the code base" 1) 30+ people work in this and it is millions of lines. 2) VIM, Emacs, and Eclipse can all do what we are asking for here, I am one of two people that use VS Code and making a huge whitespace commit for just us is not feasible when 3) some of our developers have not and seemingly will not configure their editors to use spaces instead of tabs so the whitespace commit would be ruined as soon as one of them pushed something.

**Comment by scigor** [comment]
Same usecase here as @marktucker and @Wintaru. I hope this gets higher priority, because this is the reason vscode does not get higher adoption at the company i am working at.

**Comment by Timmmm** [comment]
It's a bit confusing that both settings still exist internally in the VSCode. If I'm adding a new feature do I just use `tabSize` for everything or should I try and do things "correctly" using `indentSize` and `tabSize`?

I think having too settings is overkill and I haven't heard a compelling reason to have tab size different to indent size. Legacy code can easily be fixed with `sed` or an automatic formatter (even if it is a million lines @Wintaru), and I don't think VSCode should greatly complicate itself just to support really really unusual code styles.

**Comment by Wintaru** [comment]
> It's a bit confusing that both settings still exist internally in the VSCode. If I'm adding a new feature do I just use `tabSize` for everything or should I try and do things "correctly" using `indentSize` and `tabSize`?
> 
> I think having two settings is overkill and I haven't heard a compelling reason to have tab size different to indent size. Legacy code can easily be fixed with `sed` or an automatic formatter (even if it is a million lines @Wintaru), and I don't think VSCode should greatly complicate itself just to support really really unusual code styles.

I’m sorry you think this is confusing but Vim, Emacs, Eclipse, and other major editors support this. If it is already an internal setting why can’t we expose this to settings and let us set both? Your excuse doesn’t fly with me, sorry.

**Comment by marktucker** [comment]
Automated updating of a huge, old code base isn't impossible because the scripts are hard to write, but becuase running those scripts and committing the results erases, or at least obscures and greatly complicates access to 25 years of svn blame history as in one commit 90% of the lines in the code based are changed by these scripts. That is a price I cannot pay just to be able to use VSCode.

**Comment by Timmmm** [comment]
> I’m sorry you think this is confusing but Vim, Emacs, Eclipse, and other major editors support this.

You misunderstand. I meant it is confusing when developing VSCode itself that both settings exist internally in VSCode, but it actually only supports a single setting.

> running those scripts and committing the results erases

Clearly not.

> or at least obscures and greatly complicates access to 25 years of svn blame

Actually, just for converting all tabs to spaces you could avoid a big whitespace modification commit by rewriting history and running the script for every commit. Might take a while but git has built in support to do it automatically (`filter-branch`).

In any case, I still don't think it is worth the extra complication just to support the "I don't want to use `git blame -w`" use case.

**Comment by rmunn** [comment]
> I think having two settings is overkill and I haven't heard a compelling reason to have tab size different to indent size. Legacy code can easily be fixed with `sed` or an automatic formatter (even if it is a million lines @Wintaru), and I don't think VSCode should greatly complicate itself just to support really really unusual code styles.

The compelling reason, which may not feel compelling to you but is compelling to many people, is that many tools (including some languages) are hardcoded to treat Tab characters as being 8 spaces wide. For example, [Python 2](https://docs.python.org/2.0/ref/indentation.html) and [Python 3](https://docs.python.org/3.8/reference/lexical_analysis.html#indentation) both treat a Tab character as meaning "one to eight spaces such that the total number of characters up to and including the replacement is a multiple of eight (this is intended to be the same rule as used by Unix)". Note that Python 3 adds the rule "Indentation is rejected as inconsistent if a source file mixes tabs and spaces in a way that makes the meaning dependent on the worth of a tab in spaces; a TabError is raised in that case," which means that mixed tabs-and-spaces files will cause a syntax error in Python 3 until the codebase is fixed to be consistent. But for codebases using Python 2, where four-space indents are the norm but someone in the project has created a file containing tab characters (which is equivalent to eight spaces, i.e. two indentation levels), it's crucial that `tabSize` be left at 8 while `indentSize` needs to be 4 in order for the existing code to be displayed and edited correctly.

And the suggested "just rewrite history with `git filter-branch`" solution is a non-starter for many projects. That will change every commit hash, which would end up causing all sorts of links (GitHub PRs, mentions of commit IDs in Git history such as reverted commits) to go dead. That's far more costly than having two settings in an editor, one for tab size and one for indent size. Those two options *should* not be the same value in many projects, and a decent editor should be able to handle that use case.

**Comment by Timmmm** [comment]
Good point about links being invalidated. However I think Python's mad indentation rules are irrelevant. Nobody mixes tabs and spaces in Python code precisely because it is so hard to reason about.

**Comment by jedmao** [comment]
I haven't heard anyone yet make the distinction between indentation and alignment. Ideally, a document would [only use tab characters for indentation and spaces for alignment](https://dmitryfrank.com/articles/indent_with_tabs_align_with_spaces).

![tabs-vs-spaces](https://cloud.githubusercontent.com/assets/1058243/18036844/49a9426c-6d3b-11e6-98ae-2c9c3db623ae.gif)

If I were building an editor from scratch, I would keep the word "tab" out of the setting name and start with something like the following:

```ts
interface EditorSettings {
  /**
   * Each indent is represented by either a single tab character or a specified number of spaces.
   */
  indent?: 'tab' | number;

  /**
   * Visual width for each level of indentation.
   * @default EditorSettings['indent']
   * @example an `indent` of `2` and an `indentWidth` of `4` would make each space in the
   * indent range appear as `2` spaces of width. Conversely, an `indent` of `4` with an
   * `indentWidth` of `2` would make each space character appear at half width.
   */
  indentWidth?: number;

  /**
   * In alignment regions, each Tab key press inserts a tab character.
   * @default false
   * @warning changing the indentation width will break the alignment if this is `true`.
   */
  alignViaTabs?: boolean;

  /**
   * If `alignViaTabs` is `false`, pressing the Tab key within alignment regions produces the
   * correct number of spaces to land on a column number that is divisible by this number.
   */
  tabStops?: number;
}
```

### Related issues
- [Smart tabs - Feature Request](https://github.com/microsoft/vscode/issues/33974)
- [EditorConfig | Smart Tabs](https://github.com/editorconfig/editorconfig/issues/323)
- https://github.com/editorconfig/editorconfig-vscode/issues/190

**Comment by alexdima** [comment]
Done via https://github.com/microsoft/vscode/pull/155450
